h5netcdf/core.py

```diff
@@ -125,28 +132,62 @@ def shuffle(self):
 
 NOT_A_VARIABLE = b'This is a netCDF dimension but not a netCDF variable.'
 
+class lazy_objects():
+    def __init__(self,object_cls):
+        self._objects=OrderedDict()
+        self._loaded_objects=dict()
+        self._object_cls=object_cls
+        return
+
+    def set(self,name,object):
+        self._loaded_objects[name]=object
+        self._objects[name]=name
+
+    def keys(self):
+        return self._objects.keys()
+
+    def __iter__(self):
+        for name in self._objects:
+            yield name
+
+    def __len__(self):
+        return len(self._objects)
+
+    def __setitem__(self,key,value):
+        self._objects[key]=value
+        return
+
+    def __getitem__(self,key):
+        if key in self._loaded_objects.keys():
+            return self._loaded_objects[key]
+        else:
+            self._loaded_objects[key]=self._object_cls(self._objects[key])
+            return self[key] 
 
 class Group(Mapping):
+
     _variable_cls = Variable
 
     @property
     def _group_cls(self):
         return Group
 
-    def __init__(self, parent, h5group):
+    def __init__(self, parent, name):
         self._parent = parent
         self._root = parent._root
-        self._h5group = h5group
+        self._h5path='/'.join([parent.name.rstrip('/'),name.lstrip('/')])
+        self._lazy=parent._lazy
+
         if parent is not self:
             self._dim_sizes = parent._dim_sizes.new_child()
             self._dim_order = parent._dim_order.new_child()
 
-        self._variables = OrderedDict()
-        self._groups = OrderedDict()
+        self._variables = lazy_objects(lambda x:self._variable_cls(self,x))
+        self._groups = lazy_objects(lambda x:self._group_cls(self,x))
 
-        for k, v in h5group.items():
+        for k, v in self._h5group.items():
             if isinstance(v, h5py.Group):
-                self._groups[k] = self._group_cls(self, v)
+                self._groups[k] = k
```

Comment:
This feels redundant. I would suggest adding a `add` method to lazy_objects instead.